### ISSUE TITLE
Notepad++ ignores all API XML files in "Plugins\APIs"?

### ISSUE THREAD

**Issue by Ola-Ponny** [notepad-plus-plus/notepad-plus-plus#5149]
To make this story very, very short (I've been trying to find this out for literally *years* now...):

I'm trying to make Notepad++ recognize the PHP functions that belong to my own library. I've found the "C:\Program Files\Notepad++\plugins\APIs" dir, which contains a bunch of files such as php.xml. I copied the php.xml, ripped out all the guts and added my own function names in the same syntax style (which seems completely undocumented). I called it MyLib.xml and it's located in the same dir.

I then restart Notepad++ and type one of my function names. It's not recognized. It doesn't matter if I type it in a .php file or other kind of file, or if I select "PHP" in the "Language" list or not. There is no "MyLib" listed there. I've searched through all configs and menu items and searched and asked around and I just cannot find out how to make Notepad++ recognie my library's functions.

After countless failed experiments, I desperately tried to delete contents from the php.xml file, and even deleted it, but this is entirely ignored by Notepad++. It doesn't seem to be loading those files at all, because it has all the standard PHP functions highlighted no matter what php.xml contains, or indeed even if it's deleted entirely!

I've even tried making up the dir in %APPDATA%, just because somebody online claimed that's how it's done, but the same thing happens there: nothing.

Since this entire format and procedure is entirely undocumented... what do I do?

PS: Please note that I'm not asking about the auto-complete stuff. I know how to enable that, should I want it. But it's just annoying. I only want my functions to be hi-lit like the built-in PHP functions are (for example: "abs()") and clearly, these "API" files are ignored by Notepad++...

**Comment by DavidHansen-0x454d43** [comment]
@Ola-Ponny, is what you are trying to do is get Notepad++ to make your functions names bold and blue? 

I tried modifying PHP.xml, and it would autocomplete my new entry, but it wouldn't bold/blue it. One thing I discovered is that any malformed XML is silently ignored by Notepad++.  So if it appears like N++isn't loading your file, run it through a linter. 

I think the keywords N++ bolds/blues are defined internally by Scintilla's lexer, and isn't user configurable in Notepad++.

**Comment by Ola-Ponny** [comment]
> @Ola-Ponny, is what you are trying to do is get Notepad++ to make your functions' names bold and blue?

Yes, with the default theme, it's bold and blue indeed. Or rather, it isn't. Only for PHP built-in functions such as abs and strpos. Not my own functions, defined in MyLib.xml.

And thinking further about it, how would Notepad++ automatically know to load MyLib.xml anyway? Surely I must tell it somehow that this is what I want? Or is it supposed to parse all files in that dir? Not trying to be rude, but please re-read my post as I feel that we may be talking about slightly different things. (As often is the case when I interact with humans.)

> I tried modifying PHP.xml, and it would autocomplete my new entry, but it wouldn't bold/blue it. I think the keywords N++ bolds/blues are defined internally by Scintilla's lexer, and isn't user configurable in Notepad++. When I changed the ABS entry in PHP.XML to ABS2, typing 'A' presented the ABS2 autocomplete option and not ABS. However it didn't bold/blue ABS2, while it did still bold/blue for ABS.

Well, auto-complete is of no use to me. I want it to highlight the functions so I quickly know that I haven't made a typo (frequently happens).

I have also noticed that no matter how well the text looks in the file, and no matter what encoding I specify/use, any "special" chars are garbled in the Notepad++ list of autocomplete suggestions... but this is really getting off-topic from my original question, which has nothing to do with auto-complete. Although that seems to be what those "APIs" .xml files seem to be for... which is a huge disappointment to me. Also makes little sense IMO.

> p.s. One thing I discovered is that any malformed XML is silently ignored by Notepad++. So if it appears like N++isn't loading your file, run it through a linter.

Well, it looks perfect syntax-wise when compared to the others, and the suggestions box (as mentioned) is filled up, so it clearly is parsed in that sense, but again, I'm talking about the highlighting when looking at the code file. (Those "suggestions" drive me insane when turned on, and only cause swearing and lost productivity.)

I apologize in advance if I sound rude. I'm just incredibly frustrated and tired of undocumented nonsense and have wasted so many hours of my life on this specific problem.

(Make sure to never have the function list turned on when you load those API .xml files or Notepad++ will freeze for many seconds each time you focus the window/tab.)

**Comment by DavidHansen-0x454d43** [comment]
> And thinking further about it, how would Notepad++ automatically know to load MyLib.xml anyway? Surely I must tell it somehow that this is what I want? Or is it supposed to parse all files in that dir? Not trying to be rude, but please re-read my post as I feel that we may be talking about slightly different things. (As often is the case when I interact with humans.)

Notepad++ _doesn't_ know to load MyLib.xml. It loads whatever xml file it needs to based on what language is selected. If you want it to read MyLib.xml, you'll need to create your own language (Menu->Language->Define your language...' and then save that UDL as "MyLib". 

I don't think this is what you really want though, because if you choose your "MyLib" langauge then you lose the coloring/folding that the PHP lexer provides. So you would need to add your custom keywords into php.xml. 

When I tested this though, my modified/custom keywords in php.xml did not get colored the same as other PHP reserved words... so this led me to believe that Scintilla's PHP lexer has its own keyword list it uses to apply the blue/bold style. 

>I have also noticed that no matter how well the text looks in the file, and no matter what encoding I specify/use, any "special" chars are garbled in the Notepad++ list of autocomplete suggestions... 

I doubt you'll be able to use anything other than standard ASCII characters, unfortunately. 

>I apologize in advance if I sound rude. I'm just incredibly frustrated and tired of undocumented nonsense and have wasted so many hours of my life on this specific problem.

Nah, you didn't come off sounding rude. I've implemented my own franken-lexer using UDL/pythonscript, and I've felt the same frustration you are feeling. I'm still frustrated at the lack of documentation and the locked/dead wiki. 

The pythonscript plugin does a good job of filling in the gaps of Notepad++'s functionality. It might be the best route to achieve what your trying to do. If you'd like help, hop on over to https://notepad-plus-plus.org/community/

**Comment by Ola-Ponny** [comment]
> I doubt you'll be able to use anything other than standard ASCII characters, unfortunately.

I should reply more in general, but right now, I'll just comment that my functions are full of non-ASCII characters, as would virtually all natural non-English identifiers. Forcing only ASCII seems too archaic to be plausible in 2018... especially as Notepad++ itself very much supports UTF-8?

**Comment by chcg** [comment]
@DavidHansen-0x454d43 
@Ola-Ponny 
Did you check current docu on UDL at https://ivan-radic.github.io/udl-documentation ?
Regarding lexer/folding, see https://www.scintilla.org/ScintillaDoc.html as these are baic functions of scintilla itself.

If you want to modify the php language itself langs.xml seems to be the file which needs to be modified. Maybe in combination with stylers.xml. Depending on your installation the place on disc is different. 
The template therefore are https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/langs.model.xml and https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/stylers.model.xml.

Anyhow https://notepad-plus-plus.org/community/ is a good place to checkout previous such conversations or start a new thread.

**Comment by Ola-Ponny** [comment]
> langs.xml

Wasted another few hours editing that file (after copying langs.model.xml since langs.xml doesn't exist by default) and no matter what, it's entirely ignored by Notepad++...

**Comment by DavidHansen-0x454d43** [comment]
I'm pretty sure that N++'s XML parser doesn't handle non-ASCII characters, so updating config files isn't going to work for you. 

I have to plug PythonScript again. It handles UTF-8 characters, so it will be able to find and style your function declarations.

**Comment by sasumner** [comment]
@DavidHansen-0x454d43 .

>I have to plug PythonScript again. It handles UTF-8 characters

..as long as there aren't any non-ASCII characters in Pythonscripts path:  https://github.com/bruderstein/PythonScript/issues/112
:-)

**Comment by Ola-Ponny** [comment]
> I'm pretty sure that N++'s XML parser doesn't handle non-ASCII characters, so updating config files isn't going to work for you.

I tried it with "test12356353test" -- same thing. It's ignored.

It's difficult to describe in text how angry I am at this problem. It really looks ugly when my own library's function names and constants aren't properly highlighted... And I've looked long and hard without finding anything better than Notepad++...

**Comment by DavidHansen-0x454d43** [comment]
@Ola-Ponny, see [this thread](https://notepad-plus-plus.org/community/topic/15983/php-additional-syntax-highlighting) for some information on why this isn't working. My understanding, is that you can modify php.xml to add function names to the keyword autocomplete list, but they won't be styled and you won't be able to use non-ASCII characters. Since you function names contain non-ASCII characters, even if you could get custom PHP keywords to be styled properly, it won't do any good because your non-ASCII characters would break it. 

And like @sasumner mentioned, if you have any non-ASCII characters in your username (and hence %APPDATA% path), then "fun" unexpected things are likely to happen.
